Sources/MongoSwift/MongoConnectionString.swift

```diff
@@ -0,0 +1,197 @@
+                        in: hostAndPort,
+                        range: NSRange(hostAndPort.startIndex..<hostAndPort.endIndex, in: hostAndPort)
+                    ),
+                    let hostRange = Range(match.range(at: 1), in: hostAndPort)
+                else {
+                    throw MongoError.InvalidArgumentError(message: "couldn't parse address from \(hostAndPort)")
+                }
+                host = String(hostAndPort[hostRange])
+                if let portRange = Range(match.range(at: 2), in: hostAndPort) {
+                    port = try HostIdentifier.verifyPort(from: String(hostAndPort[portRange]), scheme: scheme)
+                } else {
+                    port = scheme == .srv ? nil : 27017
+                }
+                self.type = .ipLiteral
+            } else {
+                let parts = hostAndPort.components(separatedBy: ":")
+                host = String(parts[0])
+                guard parts.count <= 2 else {
+                    throw MongoError.InvalidArgumentError(
+                        message: "expected only a single port delimiter ':' in \(hostAndPort)"
+                    )
+                }
+                if parts.count > 1 {
+                    port = try HostIdentifier.verifyPort(from: parts[1], scheme: scheme)
+                } else {
+                    port = scheme == .srv ? nil : 27017
+                }
+                self.type = .hostname
+            }
+            self.host = host
+            self.port = port
+        }
+
+        public var description: String {
+            switch self.type {
+            case .ipLiteral:
+                if let prt = port {
+                    return "[\(self.host)]:\(prt)"
+                } else {
+                    return "[\(self.host)]"
+                }
+            default:
+                if let prt = port {
+                    return "\(self.host):\(prt)"
+                } else {
+                    return "\(self.host)"
+                }
+            }
+        }
+    }
+
+    /// Parses a new `MongoConnectionString` instance from the provided string.
+    /// - Throws:
+    ///   - `MongoError.InvalidArgumentError` if the input is invalid.
+    public init(throwsIfInvalid input: String) throws {
+        let schemeAndRest = input.components(separatedBy: "://")
+        guard schemeAndRest.count == 2, let scheme = Scheme(schemeAndRest[0]) else {
+            throw MongoError.InvalidArgumentError(
+                message: "Invalid connection string scheme, expecting \'mongodb\' or \'mongodb+srv\'"
+            )
+        }
+        guard !schemeAndRest[1].isEmpty else {
+            throw MongoError.InvalidArgumentError(message: "Invalid connection string")
+        }
+        let identifiersAndOptions = schemeAndRest[1].components(separatedBy: "/")
+        guard identifiersAndOptions.count <= 2 else {
+            throw MongoError.InvalidArgumentError(message: "Host with unescaped slash")
```

Comment:
I don't think this error message quite lines up with the condition it's testing. For instance, the following connection string will yield the error, but the problem isn't really with the host identifier:
```
mongodb://localhost:27017/?hello=world/dog=cat
```
That being said, I think it's more the scope of SWIFT-1174 to improve this, so I'm fine with just leaving it and adding a `TODO` with the ticket number so we remember to improve it.

Reply:
done